### PR TITLE
added number_of_islands_dfs algo

### PR DESCRIPTION
**Pull request by harshagrawal523**
#### Description of Change

Added the algorithm for Number of Islands questions using DFS

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

**Comment by harshagrawal523** [comment]
@mishraabhinn can you please approve the PR or add the hacktoberfest tag on it.

**Comment by harshagrawal523** [comment]
> @mishraabhinn can you please approve the PR or add the hacktoberfest tag on it.

@mishraabhinn please review it

File: graph/number_of_islands_dfs.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <map>
+#include <numeric>
+#include <set>
+#include <string>
+#include <unordered_map>
+#include <vector>
+//#include <bits/stdc++.h>
```
Comment (mishraabhinn) [inline graph/number_of_islands_dfs.cpp:10]: Remove the dead code

File: graph/number_of_islands_dfs.cpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+using namespace std;
```
Comment (mishraabhinn) [inline graph/number_of_islands_dfs.cpp:11]: Do not use this

File: graph/number_of_islands_dfs.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    dfs(matrix, row - 1, col);
+    dfs(matrix, row + 1, col);
+    dfs(matrix, row, col + 1);
+    dfs(matrix, row, col - 1);
+}
+
+int32_t main() {
+    ios_base::sync_with_stdio(false);
+    cin.tie(0);
+    cout.tie(0);
```
Comment (mishraabhinn) [inline graph/number_of_islands_dfs.cpp:32]: This is not competitive programing. So do not use this here 😄

File: graph/number_of_islands_dfs.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    }
+    for (int i = 0; i < n; i++) {
+        for (int j = 0; j < m; j++) {
+            if (matrix[i][j] == '1') {
+                dfs(matrix, i, j);
+                ans++;
+            }
+        }
+    }
+
```
Comment (mishraabhinn) [inline graph/number_of_islands_dfs.cpp:55]: Need some automatic tests. Could you please do that?

File: graph/number_of_islands_dfs.cpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <map>
+#include <numeric>
+#include <set>
+#include <string>
+#include <unordered_map>
+#include <vector>
```
Comment (mishraabhinn) [inline graph/number_of_islands_dfs.cpp:9]: Please read contributing guidelines and add relevant comments to your code.

**Review by mishraabhinn — CHANGES_REQUESTED** [review]

File: graph/number_of_islands_dfs.cpp (hunk 4 above)
Comment (harshagrawal523) [reply graph/number_of_islands_dfs.cpp:55]: Can you explain what are automatic test?

File: graph/number_of_islands_dfs.cpp (hunk 2 above)
Comment (harshagrawal523) [reply graph/number_of_islands_dfs.cpp:10]: Removed this in recent PR

File: graph/number_of_islands_dfs.cpp (hunk 3 above)
Comment (harshagrawal523) [reply graph/number_of_islands_dfs.cpp]: Okay 😅

File: graph/number_of_islands_dfs.cpp (hunk 4 above)
Comment (mishraabhinn) [reply graph/number_of_islands_dfs.cpp]:
```
/**
 * @brief Self-test implementations
 * @returns void
 */
static void test() {
    /* descriptions of the following test */
    assert(func(...) == ...); // this ensures that the algorithm works as expected

    // can have multiple checks
}
```

Comment (harshagrawal523) [reply graph/number_of_islands_dfs.cpp]: I have added the appropriate changes in the recent PR please check once.

**Comment by harshagrawal523** [comment]
@Panquesito7  why is Awesome CI Workflow / Code Formatter (pull_request)  failing?

File: graph/number_of_islands_dfs.cpp [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+ *
+ * The algorithm implemented in this file determines the number of islands in a
+ * 2D array of 1s(land) and 0s(water).
+ *
+ * @author [Harsh Agrawal](https://github.com/harshagrawal523)
+ *
+ */
+#include <iostream>
+#include <string>
+#include <vector>
```
Comment (Panquesito7) [inline graph/number_of_islands_dfs.cpp:21]: Please add a one-line description of what the library/header is for (**see the example below**).

```c++
#include <iostream>    /// for IO operations
#include <cassert>    /// for assert
```

File: graph/number_of_islands_dfs.cpp [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+/**
+ * @namespace islands
+ * @brief DFS algorithms
+ */
+namespace islands {
```
Comment (Panquesito7) [inline graph/number_of_islands_dfs.cpp:27]: This should be `graph`, IMO.
You will need to change other instances to `graph`.
```suggestion
/**
 * @namespace graph
 * @brief Graphical algorithms
 */
namespace graph {
```

File: graph/number_of_islands_dfs.cpp [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+ * @namespace number_of_islands
+ * @brief Functions for counting the number of islands
+ */
+namespace number_of_islands {
+/**
+ * @brief Class for representing the matrix of 0s(water) and 1s(land).
+ */
+class matrix {
+ private:
+    int n, m;  // dimensions of graph
```
Comment (Panquesito7) [inline graph/number_of_islands_dfs.cpp:38]:
```suggestion
    int n, m;  ///< dimensions of the graph
```

File: graph/number_of_islands_dfs.cpp [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    std::vector<std::vector<char> >
+        mat;  // matrix that stores the ocean(0s and 1s)
```
Comment (Panquesito7) [inline graph/number_of_islands_dfs.cpp:40]:
```suggestion
        mat;  ///< matrix that stores the ocean(0s and 1s)
```

File: graph/number_of_islands_dfs.cpp [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    /**
+     * @brief Constructor that initializes the matrix of oceans on creation
+     */
+    explicit matrix(std::vector<std::vector<char> > mat, int rows, int cols) {
+        n = rows;
+        m = cols;
+        mat.resize(n, std::vector<char>(m));
+    }
+    void dfs(std::vector<std::vector<char> > &mat, int row,
+             int col);  // dfs function that helps us find the number of islands
```
Comment (Panquesito7) [inline graph/number_of_islands_dfs.cpp:52]:
```suggestion
             int col);  ///< DFS function that helps us find the number of islands
```

File: graph/number_of_islands_dfs.cpp [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    dfs(matrix, row - 1, col);
+    dfs(matrix, row + 1, col);
+    dfs(matrix, row, col + 1);
+    dfs(matrix, row, col - 1);
+}
+
+}  // namespace number_of_islands
+}  // namespace islands
+
+static void test() {
```
Comment (Panquesito7) [inline graph/number_of_islands_dfs.cpp:75]:
```suggestion
/**
 * @brief Self-test implementations
 * @returns void
 */
static void test() {
```

File: graph/number_of_islands_dfs.cpp [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+        }
+    }
+    std::cout << ans << std::endl;
+}
+
+int main() {
+    test();  // Testing
+
+    return 0;
+}
```
Comment (Panquesito7) [inline graph/number_of_islands_dfs.cpp:103]: Missing an endline. Without this endline, there could be issues with some IDEs and configurations.
```suggestion
}

```

File: graph/number_of_islands_dfs.cpp [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+            if (mat[i][j] == '1') {
+                ocean.dfs(mat, i, j);
+                ans++;
+            }
+        }
+    }
+    std::cout << ans << std::endl;
+}
+
+int main() {
```
Comment (Panquesito7) [inline graph/number_of_islands_dfs.cpp:99]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
int main() {
```

File: graph/number_of_islands_dfs.cpp [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        mat.resize(n, std::vector<char>(m));
+    }
+    void dfs(std::vector<std::vector<char> > &mat, int row,
+             int col);  // dfs function that helps us find the number of islands
+};
+/**
+ * @brief this is a dfs funtion which checks all the connected 1s
+ * the function returns if it goes out of bound of matrix
+ */
+void matrix::dfs(std::vector<std::vector<char> > &matrix, int row, int col) {
```
Comment (Panquesito7) [inline graph/number_of_islands_dfs.cpp:58]: Here seems to be the issue. Check the CI logs https://github.com/TheAlgorithms/C-Plus-Plus/actions/runs/3169823215/jobs/5162010561#step:10:47

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 🚀
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/harshagrawal523/C-Plus-Plus/actions

File: graph/number_of_islands_dfs.cpp (hunk 14 above)
Comment (harshagrawal523) [reply graph/number_of_islands_dfs.cpp]: I guess I have resolved this issue but something is still wrong and I have no idea what is wrong

Comment (harshagrawal523) [reply graph/number_of_islands_dfs.cpp]: Can you check what's the issue? @Panquesito7 @mishraabhinn

**Comment by harshagrawal523** [comment]
@Panquesito7 @mishraabhinn  can you please review the PR?

**Comment by harshagrawal523** [comment]
Can you please accept the PR @mishraabhinn @Panquesito7

**Comment by harshagrawal523** [comment]
@Panquesito7 why is it showing a workflow run error?

**Comment by mishraabhinn** [comment]
> @Panquesito7 why is it showing a workflow run error?
Screenshot 2022-10-09 at 12 29 50

**Comment by harshagrawal523** [comment]
> > @Panquesito7 why is it showing a workflow run error?
> > Screenshot 2022-10-09 at 12 29 50

I have removed this error in the latest PR , still there is some issue which makes the test fail everytime. Can you please help in that?

File: graph/number_of_islands_dfs.cpp [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+ *
+ * The algorithm implemented in this file determines the number of islands in a
+ * 2D array of 1s(land) and 0s(water).
+ *
+ * @author [Harsh Agrawal](https://github.com/harshagrawal523)
+ *
+ */
+#include <iostream>  // for IO operations
+#include <string>    // for strings
+#include <vector>    // for vectors
```
Comment (Panquesito7) [inline graph/number_of_islands_dfs.cpp:21]:
```suggestion
#include <iostream>  /// for IO operations
#include <string>    /// for string manipulation
#include <vector>    /// for std::vector
```

File: graph/number_of_islands_dfs.cpp [hunk 16]
```diff
@@ -0,0 +1,113 @@
+/**
+ * @file
+ *
+ * @brief Algorithm to count the number of islands
```
Comment (Panquesito7) [inline graph/number_of_islands_dfs.cpp:4]: Provide a Wikipedia link in Markdown format (if available/possible).
If there's no Wikipedia link, add another web source for algorithm explanation. 🙂

File: graph/number_of_islands_dfs.cpp [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+ * @namespace number_of_islands
+ * @brief Functions for counting the number of islands
+ */
+namespace number_of_islands {
+/**
+ * @brief Class for representing the matrix of 0s(water) and 1s(land).
+ */
+class matrix {
+ private:
+    int n, m;  ///< dimensions of the graph
```
Comment (Panquesito7) [inline graph/number_of_islands_dfs.cpp:38]: Shouldn't we initialize this to `0`?
```suggestion
    int n = 0, m = 0;  ///< dimensions of the graph
```

File: graph/number_of_islands_dfs.cpp [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+
+}  // namespace number_of_islands
+}  // namespace graph
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+
+static void test() {
```
Comment (Panquesito7) [inline graph/number_of_islands_dfs.cpp:81]: Self-test implementation checks should contain `assert`.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: graph/number_of_islands_dfs.cpp [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+    explicit matrix(std::vector<std::vector<char> > mat, int rows, int cols) {
+        n = rows;
+        m = cols;
+        mat.resize(n, std::vector<char>(m));
+    }
+    void dfs(
+        std::vector<std::vector<char> > *mat, int row,
+        int col);  ///< DFS function that helps us find the number of islands
+};
+/**
```
Comment (krishnacx) [inline graph/number_of_islands_dfs.cpp:55]: Add empty line.

**Pull request closed without merging**